db/version_set.cc

```diff
@@ -330,94 +330,69 @@ Status Version::Get(const ReadOptions& options, const LookupKey& k,
-      files = &tmp[0];
-      num_files = tmp.size();
-    } else {
-      // Binary search to find earliest index whose largest key >= ikey.
-      uint32_t index = FindFile(vset_->icmp_, files_[level], ikey);
-      if (index >= num_files) {
-        files = nullptr;
-        num_files = 0;
-      } else {
-        tmp2 = files[index];
-        if (ucmp->Compare(user_key, tmp2->smallest.user_key()) < 0) {
-          // All of "tmp2" is past any data for user_key
-          files = nullptr;
-          num_files = 0;
-        } else {
-          files = &tmp2;
-          num_files = 1;
-        }
-      }
-    }
+    static bool Match(void* arg, int level, FileMetaData* f) {
+      State* state = reinterpret_cast<State*>(arg);
 
-    for (uint32_t i = 0; i < num_files; ++i) {
-      if (last_file_read != nullptr && stats->seek_file == nullptr) {
+      if (state->stats->seek_file == nullptr) {
         // We have had more than one seek for this read.  Charge the 1st file.
-        stats->seek_file = last_file_read;
-        stats->seek_file_level = last_file_read_level;
+        state->stats->seek_file = f;
+        state->stats->seek_file_level = level;
       }
 
-      FileMetaData* f = files[i];
-      last_file_read = f;
-      last_file_read_level = level;
-
       Saver saver;
       saver.state = kNotFound;
-      saver.ucmp = ucmp;
-      saver.user_key = user_key;
-      saver.value = value;
-      s = vset_->table_cache_->Get(options, f->number, f->file_size, ikey,
-                                   &saver, SaveValue);
+      saver.ucmp = state->ucmp;
+      saver.user_key = state->user_key;
+      saver.value = state->value;
+
+      Status s = state->vset->table_cache_->Get(*state->options, f->number,
+                                                f->file_size, state->ikey,
+                                                &saver, SaveValue);
       if (!s.ok()) {
-        return s;
+        state->s = s;
+        return false;
       }
       switch (saver.state) {
         case kNotFound:
-          break;  // Keep searching in other files
+          return true; // Keep saerching in other files
         case kFound:
-          return s;
+          state->s = s;
+          return false;
         case kDeleted:
-          s = Status::NotFound(Slice());  // Use empty error message for speed
-          return s;
+          return false;
         case kCorrupt:
-          s = Status::Corruption("corrupted key for ", user_key);
-          return s;
+          state->s = Status::Corruption("corrupted key for ", state->user_key);
+          return false;
       }
     }
-  }
+  };
+
+  stats->seek_file = nullptr;
+  stats->seek_file_level = -1;
+
+  State state;
+  state.s = Status::NotFound(Slice());
+  state.stats = stats;
+  state.ikey = ikey;
+  state.user_key = user_key;
+  state.ucmp = ucmp;
+  state.value = value;
+  state.vset = vset_;
+
+  ForEachOverlapping(user_key, ikey, &state, &State::Match);
```

Comment:
Get() can be performance sensitive.  Have you run db_bench with full optimizations on before and after this change?

Reply:
Before:
LevelDB:    version 1.22
Keys:       16 bytes each
Values:     100 bytes each (50 bytes after compression)
Entries:    1000000
RawSize:    110.6 MB (estimated)
FileSize:   62.9 MB (estimated)
------------------------------------------------
fillseq      :       8.881 micros/op;   12.5 MB/s
fillsync     :    3992.566 micros/op;    0.0 MB/s (1000 ops)
fillrandom   :      11.409 micros/op;    9.7 MB/s
overwrite    :      12.871 micros/op;    8.6 MB/s
readrandom   :       6.165 micros/op; (1000000 of 1000000 found)
readrandom   :       5.795 micros/op; (1000000 of 1000000 found)
readseq      :       0.261 micros/op;  423.9 MB/s
readreverse  :       0.467 micros/op;  236.9 MB/s
compact      : 1089242.000 micros/op;
readrandom   :       4.761 micros/op; (1000000 of 1000000 found)
readseq      :       0.311 micros/op;  355.3 MB/s
readreverse  :       0.475 micros/op;  232.8 MB/s
fill100K     :    1422.327 micros/op;   67.1 MB/s (1000 ops)
crc32c       :       1.551 micros/op; 2519.0 MB/s (4K per op)

After
LevelDB:    version 1.22
Keys:       16 bytes each
Values:     100 bytes each (50 bytes after compression)
Entries:    1000000
RawSize:    110.6 MB (estimated)
FileSize:   62.9 MB (estimated)
------------------------------------------------
fillseq      :       8.603 micros/op;   12.9 MB/s
fillsync     :    3915.398 micros/op;    0.0 MB/s (1000 ops)
fillrandom   :      11.243 micros/op;    9.8 MB/s
overwrite    :      12.335 micros/op;    9.0 MB/s
readrandom   :       6.263 micros/op; (1000000 of 1000000 found)
readrandom   :       5.693 micros/op; (1000000 of 1000000 found)
readseq      :       0.236 micros/op;  468.9 MB/s
readreverse  :       0.399 micros/op;  277.2 MB/s
compact      : 1011847.000 micros/op;
readrandom   :       4.582 micros/op; (1000000 of 1000000 found)
readseq      :       0.203 micros/op;  544.6 MB/s
readreverse  :       0.309 micros/op;  358.4 MB/s
fill100K     :    1386.870 micros/op;   68.8 MB/s (1000 ops)
crc32c       :       1.349 micros/op; 2895.0 MB/s (4K per op)
snappycomp   :       3.847 micros/op; 1015.5 MB/s (output: 55.1%)
snappyuncomp :       0.613 micros/op; 6372.0 MB/s
snappycomp   :       4.281 micros/op;  912.4 MB/s (output: 55.1%)
snappyuncomp :       0.645 micros/op; 6056.2 MB/s